### ISSUE TITLE
[bug] RSS feed validation issues

### ISSUE THREAD

**Issue by mirabilos** [superseriousbusiness/gotosocial#2282]
### Describe the bug with a clear and concise description of what the bug is.

Low-priority thing: the W3C feed validator `https://validator.w3.org/feed/check.cgi?url=…` complains about the feeds produced by GtS.

## author tag

The item `author` tag’s content MUST be an eMail address.

The recommendation is…

> For a weblog authored by a single individual it would make sense to omit the `<author>` element.

… to just drop it.

## Missing source attribute: url

The `source` tag…

> value is the name of the RSS channel that the item came from

… which is obviously not the case for a GtS post feed… well, not normally. It especially makes no sense here. This is used only in feed aggregators.

Recommendation: remove the tag from the feed

## Image title doesn't match channel title

tbh I’m puzzled a bit about it, but the spec says…

> Note, in practice the image `<title>` and `<link>` should have the same value as the channel's `<title>` and `<link>`.

… which is totally redundant, but they are required sub-tags, so there.

Recommendation: make them identical, even if for the image title it does make more sense to say avatar than posts.

## Invalid HTML

This is… interesting. What we have is:

```
[…]
      <description>@mirabilos@toot.mirbsd.org made a new post: &#34;What the everlasting …?&#xA;&#xA;```&#xA;$ cat n.cc&#xA;#include &lt;stddef.h&gt;&#xA;&#xA;int main(void) {&#xA;        return (NULL == nullptr);&#xA;}&#xA;$ clang++ -Wformat -Wall -Wextra -std=c++11 -Werror -fno-lto  n.cc &#xA;n.cc:4:15: error: compa...</description>
     <content:encoded><![CDATA[<p>What the everlasting …?</p><pre><code>$ cat n.cc
#include &lt;stddef.h&gt;
[…]
```

What seems to happen here is that GtS truncates the content for the `description` tag but doesn’t ensure the result (before XML-encoding) is valid HTML. The full content is then in a nōn-standard tag.

My **personal** recommendation from over a decade of running feeds generated from both PHP and shell scripts would be to…
1. drop the entire `content:encoded` tag, and…
2. put the toot (as CDATA or properly XML-encoded) into a `<description xml:space="preserve">` tag

The `xml:space` attribute is required so that usage of the `pre` tag *inside* the toot is not broken by intermediate XML processors. (We recently had a case where a Debian RSS feed was reported broken in some feedreaders for that.)

See [MirBSD news.rss](http://www.mirbsd.org/news.rss) for an example of such a feed. (You could even put a `category` tag for each hashtag, but that’s entirely optional.)

## Missing atom:link with rel="self"

Just do what [it](https://validator.w3.org/feed/docs/warning/MissingAtomSelfLink.html) says to put a ✔ into that ☐ ☻

## `title`…?

```
      <title>[The story of Félicette, the first cat in space](https://starwalk.space/en/news/first-cat-in-space-felicette)&#xA;&#xA;Both glad 🐈‍⬛😻...</title>
```

Hmmhmmhmm. Yeah, you gotta have a title. Perhaps make it the first line of the toot if multiline, otherwise the first couple of words and an `…` ellipsis, but (Markdown/HTML) formatting entirely stripped.

### What's your GoToSocial Version?

0.12.0-rc2 git-805c67b

### GoToSocial Arch

amd64 binary

### Anything else we need to know?

As I said, this is no bother, maybe an idle task, or something to exemplarily onboard a new contributor or team member, or whatever. I just wondered about the RSS feeds.

**Comment by pnwmatt** [comment]
I'm looking into this.  Here's my current write up.

I'm proposing splitting this into multiple PRs based on complexity and compartmentalization.

Regarding https://github.com/superseriousbusiness/gotosocial/issues/3540 - it looks like @tsmethurst was working on this as recently as 2 weeks ago.  I won't scope out any changes on that since it already has their eyes on it.

## go.pkg/gorilla/feeds
GtS uses the [gorilla/feeds](https://pkg.go.dev/github.com/gorilla/feeds@v1.2.0#section-readme) package for generating the feed.

They have an existing issue open for the atom:link rel="self": https://github.com/gorilla/feeds/issues/111

## In a new pull request: (not yet begun)
1. Remove author tag
2. Remove source tag
3. Image title doesn't match channel title
    - make them identical, even if for the image title it does make more sense to say avatar than posts.
4. Invalid HTML
    - This might be a `gorilla` problem - if so, open a ticket with them
    - Remove the wrapping tag `<content:encoded>`, leaving the child content in tact
    - add the `xml:space="preserve"` parameter to the post's `<description>` tag
5. Hashtag -> Category tag
    - This might be blocked on `gorilla` - if so, open a ticket with them
    - Research if there's any formatting for this per other implementations (server and client rendering):
      - do I include the # symbol (yes?)
      - do I maintain case (don't lowercase it - let it be as stored?)
    - Add a `<category>` tag for each hashtag for each post.

## In a separate pull request: (not yet begun)
1. title tag
     - Find the max length of a title per RSS spec ($maxTitleLen)
     - Investigate how emoji impact max length calculations in an RSS feed
     - Strip markdown/html formatting from message content
       - If multiline, just use first line
       - If length is too long, truncate to `($maxTitleLen - 1) characters` and append an [ellipsis](https://www.compart.com/en/unicode/U+2026) (single unicode character)

**Comment by daenney** [comment]
We've migrated to Codeberg. Your issue has been imported as https://codeberg.org/superseriousbusiness/gotosocial/issues/2282.
